napari/_vispy/visuals/markers.py

```diff
@@ -1,10 +1,31 @@
 from vispy.scene.visuals import Markers as BaseMarkers
 
+clamp_shader = """
+float clamped_size = clamp($v_size, $canvas_size_min, $canvas_size_max);
+float clamped_ratio = clamped_size / $v_size;
+$v_size = clamped_size;
+v_edgewidth = v_edgewidth * clamped_ratio;
+gl_PointSize = $v_size + 4. * (v_edgewidth + 1.5 * u_antialias);
+"""
+
+old_vshader = BaseMarkers._shaders['vertex']
+new_vshader = old_vshader[:-2] + clamp_shader + '\n}'  # very ugly...
```

Comment:
Optional: in addition (or maybe instead) to pointing out the ugliness here, it might be useful to explain what `old_vshader[:-2]` is intended to be. I think you're trying to remove the last two lines from the original shader (i.e. `'\n}'`, insert this extra snippet, then close out the shader with the `'\n'`. You might also want to check/assert that those last two lines are what you think they are to avoid confusing errors in the future.